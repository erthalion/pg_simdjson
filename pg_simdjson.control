# pg_simdjson extension
comment = 'Parse json using simdjson library'
default_version = '0.1'
module_pathname = '$libdir/pg_simdjson'
relocatable = true
