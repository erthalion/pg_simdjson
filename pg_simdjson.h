extern "C" {
	#include <postgres.h>
	#include <fmgr.h>
	#include <utils/jsonb.h>
	#include "utils/numeric.h"
	#include "utils/fmgrprotos.h"

	#ifdef PG_MODULE_MAGIC
		PG_MODULE_MAGIC;
	#endif

	PG_FUNCTION_INFO_V1(pg_simdjson_parse);
	Datum pg_simdjson_parse(PG_FUNCTION_ARGS);

	typedef struct JsonbInState
	{
		JsonbParseState *parseState;
		JsonbValue *res;
	} JsonbInState;

	Jsonb *JsonbValueToJsonb(JsonbValue *val);

	const char *Errors[] = {
		"No errors",
		"No errors and buffer still has more data",
		"This ParsedJson can't support a document that big",
		"Error allocating memory, we're most likely out of memory",
		"Something went wrong while writing to the tape",
		"Document exceeds the user-specified depth limitation",
		"Problem while parsing a string",
		"Problem while parsing an atom starting with the letter 't'",
		"Problem while parsing an atom starting with the letter 'f'",
		"Problem while parsing an atom starting with the letter 'n'",
		"Problem while parsing a number",
		"The input is not valid UTF-8",
		"Unitialized",
		"Empty",
		"Within strings, some characters must be escapted, we "
							"found unescapted characters",
		"Missing quote at the end",
		"Unexpected error, consider reporting this problem as "
							 "you may have found a bug in simdjson",
	};
}
