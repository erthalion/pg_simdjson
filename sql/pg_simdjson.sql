CREATE EXTENSION pg_simdjson;

-- Strings.
SELECT simdjson_parse('""');				-- OK.
SELECT simdjson_parse($$''$$);			-- ERROR, single quotes are not allowed
SELECT simdjson_parse('"abc"');			-- OK
SELECT simdjson_parse('"abc');			-- ERROR, quotes not closed
SELECT simdjson_parse('"abc
def"');				-- ERROR, unescaped newline in string constant
SELECT simdjson_parse('"\n\"\\"');		-- OK, legal escapes
SELECT simdjson_parse('"\v"');			-- ERROR, not a valid JSON escape
-- see json_encoding test for input with unicode escapes

-- Numbers.
SELECT simdjson_parse('1');				-- OK
SELECT simdjson_parse('0');				-- OK
SELECT simdjson_parse('01');				-- ERROR, not valid according to JSON spec
SELECT simdjson_parse('0.1');				-- OK
SELECT simdjson_parse('9223372036854775808');	-- OK, even though it's too large for int8
SELECT simdjson_parse('1e100');			-- OK
SELECT simdjson_parse('1.3e100');			-- OK
SELECT simdjson_parse('1f2');				-- ERROR
SELECT simdjson_parse('0.x1');			-- ERROR
SELECT simdjson_parse('1.3ex100');		-- ERROR

-- Arrays.
SELECT simdjson_parse('[]');				-- OK
SELECT simdjson_parse('[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]');  -- OK
SELECT simdjson_parse('[1,2]');			-- OK
SELECT simdjson_parse('[1,2,]');			-- ERROR, trailing comma
SELECT simdjson_parse('[1,2');			-- ERROR, no closing bracket
SELECT simdjson_parse('[1,[2]');			-- ERROR, no closing bracket

-- Objects.
SELECT simdjson_parse('{}');				-- OK
SELECT simdjson_parse('{"abc"}');			-- ERROR, no value
SELECT simdjson_parse('{"abc":1}');		-- OK
SELECT simdjson_parse('{1:"abc"}');		-- ERROR, keys must be strings
SELECT simdjson_parse('{"abc",1}');		-- ERROR, wrong separator
SELECT simdjson_parse('{"abc"=1}');		-- ERROR, totally wrong separator
SELECT simdjson_parse('{"abc"::1}');		-- ERROR, another wrong separator
SELECT simdjson_parse('{"abc":1,"def":2,"ghi":[3,4],"hij":{"klm":5,"nop":[6]}}'); -- OK
SELECT simdjson_parse('{"abc":1:2}');		-- ERROR, colon in wrong spot
SELECT simdjson_parse('{"abc":1,3}');		-- ERROR, no value

-- Recursion.
SET max_stack_depth = '100kB';
SELECT simdjson_parse(repeat('[', 10000));
SELECT simdjson_parse(repeat('{"a":', 10000));
RESET max_stack_depth;

-- Miscellaneous stuff.
SELECT simdjson_parse('true');			-- OK
SELECT simdjson_parse('false');			-- OK
SELECT simdjson_parse('null');			-- OK
SELECT simdjson_parse(' true ');			-- OK, even with extra whitespace
SELECT simdjson_parse('true false');		-- ERROR, too many values
SELECT simdjson_parse('true, false');		-- ERROR, too many values
SELECT simdjson_parse('truf');			-- ERROR, not a keyword
SELECT simdjson_parse('trues');			-- ERROR, not a keyword
SELECT simdjson_parse('');				-- ERROR, no value
SELECT simdjson_parse('    ');			-- ERROR, no value
