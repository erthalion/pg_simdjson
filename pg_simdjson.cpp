#include "simdjson.h"
#include "simdjson.cpp"

#include "pg_simdjson.h"


void toJsonb(simdjson::ParsedJson::Iterator &pjh, JsonbInState *result,
			 JsonbIteratorToken tok)
{
	JsonbValue	v;

	if (pjh.is_object())
	{
		result->res = pushJsonbValue(&result->parseState,
									 WJB_BEGIN_OBJECT, NULL);

		if (pjh.down()) {
			/* Push key */
			v.type = jbvString;
			v.val.string.len = pjh.get_string_length();
			v.val.string.val = (char *) pjh.get_string();
			result->res = pushJsonbValue(&result->parseState, WJB_KEY, &v);

			pjh.next();
			toJsonb(pjh, result, WJB_VALUE);

			while (pjh.next())
			{
				v.type = jbvString;
				v.val.string.len = pjh.get_string_length();
				v.val.string.val = (char *) pjh.get_string();
				result->res = pushJsonbValue(&result->parseState, WJB_KEY, &v);

				pjh.next();
				toJsonb(pjh, result, WJB_VALUE);
			}
			pjh.up();
		}
		result->res = pushJsonbValue(&result->parseState,
									 WJB_END_OBJECT, NULL);
	}
	else if (pjh.is_array())
	{
		result->res = pushJsonbValue(&result->parseState,
									 WJB_BEGIN_ARRAY, NULL);
		if (pjh.down())
		{
			toJsonb(pjh, result, WJB_ELEM);

			while (pjh.next()) {
				toJsonb(pjh, result, WJB_ELEM);
			}

			pjh.up();
		}
		result->res = pushJsonbValue(&result->parseState,
									 WJB_END_ARRAY, NULL);
	}
	else
	{
		if (pjh.is_true())
		{
			v.type = jbvBool;
			v.val.boolean = true;
		}
		else if (pjh.is_false())
		{
			v.type = jbvBool;
			v.val.boolean = false;
		}
		else if (pjh.is_double())
		{
			Datum d = Float8GetDatum(pjh.get_double());

			v.type = jbvNumeric;
			v.val.numeric = DatumGetNumeric(DirectFunctionCall1(float8_numeric, d));
		}
		else if (pjh.is_integer())
		{
			Datum d = Int32GetDatum(pjh.get_integer());

			v.type = jbvNumeric;
			v.val.numeric = DatumGetNumeric(DirectFunctionCall1(int4_numeric, d));
		}
		else if (pjh.is_string())
		{
			v.type = jbvString;
			v.val.string.val = (char *) pjh.get_string();
			v.val.string.len = pjh.get_string_length();
		}
		else if (pjh.is_null())
		{
			v.type = jbvNull;
		}

		if (tok == WJB_DONE)
		{
			JsonbValue	va;

			va.type = jbvArray;
			va.val.array.rawScalar = true;
			va.val.array.nElems = 1;

			result->res = pushJsonbValue(&result->parseState, WJB_BEGIN_ARRAY, &va);
			result->res = pushJsonbValue(&result->parseState, WJB_ELEM, &v);
			result->res = pushJsonbValue(&result->parseState, WJB_END_ARRAY, NULL);
		}
		else
			result->res = pushJsonbValue(&result->parseState, tok, &v);
	}
}


Datum
pg_simdjson_parse(PG_FUNCTION_ARGS)
{
	text *json_as_text = PG_GETARG_TEXT_PP(0);

	simdjson::padded_string json(
		VARDATA_ANY(json_as_text),
		VARSIZE_ANY_EXHDR(json_as_text)
	);

	simdjson::ParsedJson pj = simdjson::build_parsed_json(json);

	if(!pj.is_valid())
	{
		ereport(ERROR,
				(errcode(ERRCODE_EXTERNAL_ROUTINE_EXCEPTION),
				 errmsg("Parsed status: not valid, %s", Errors[pj.error_code])));
	}
	else
	{
		JsonbInState result;
		simdjson::ParsedJson::Iterator pjh(pj);
		if (!pjh.is_ok())
		{
			ereport(ERROR,
				(errcode(ERRCODE_EXTERNAL_ROUTINE_EXCEPTION),
				 errmsg("Cannot create an iterator")));
		}

		memset(&result, 0, sizeof(JsonbInState));
		toJsonb(pjh, &result, WJB_DONE);

		PG_RETURN_POINTER(JsonbValueToJsonb(result.res));
	}
}
