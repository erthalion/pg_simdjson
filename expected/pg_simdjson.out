CREATE EXTENSION pg_simdjson;
-- Strings.
SELECT simdjson_parse('""');				-- OK.
 simdjson_parse 
----------------
 ""
(1 row)

SELECT simdjson_parse($$''$$);			-- ERROR, single quotes are not allowed
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('"abc"');			-- OK
 simdjson_parse 
----------------
 "abc"
(1 row)

SELECT simdjson_parse('"abc');			-- ERROR, quotes not closed
ERROR:  Parsed status: not valid, Missing quote at the end
SELECT simdjson_parse('"abc
def"');				-- ERROR, unescaped newline in string constant
ERROR:  Parsed status: not valid, Within strings, some characters must be escapted, we found unescapted characters
SELECT simdjson_parse('"\n\"\\"');		-- OK, legal escapes
 simdjson_parse 
----------------
 "\n\"\\"
(1 row)

SELECT simdjson_parse('"\v"');			-- ERROR, not a valid JSON escape
ERROR:  Parsed status: not valid, Problem while parsing a string
-- see json_encoding test for input with unicode escapes
-- Numbers.
SELECT simdjson_parse('1');				-- OK
 simdjson_parse 
----------------
 1
(1 row)

SELECT simdjson_parse('0');				-- OK
 simdjson_parse 
----------------
 0
(1 row)

SELECT simdjson_parse('01');				-- ERROR, not valid according to JSON spec
ERROR:  Parsed status: not valid, Problem while parsing a number
SELECT simdjson_parse('0.1');				-- OK
 simdjson_parse 
----------------
 0.1
(1 row)

SELECT simdjson_parse('9223372036854775808');	-- OK, even though it's too large for int8
   simdjson_parse    
---------------------
 9223372036854775808
(1 row)

SELECT simdjson_parse('1e100');			-- OK
                                            simdjson_parse                                             
-------------------------------------------------------------------------------------------------------
 10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
(1 row)

SELECT simdjson_parse('1.3e100');			-- OK
                                            simdjson_parse                                             
-------------------------------------------------------------------------------------------------------
 13000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
(1 row)

SELECT simdjson_parse('1f2');				-- ERROR
ERROR:  Parsed status: not valid, Problem while parsing a number
SELECT simdjson_parse('0.x1');			-- ERROR
ERROR:  Parsed status: not valid, Problem while parsing a number
SELECT simdjson_parse('1.3ex100');		-- ERROR
ERROR:  Parsed status: not valid, Problem while parsing a number
-- Arrays.
SELECT simdjson_parse('[]');				-- OK
 simdjson_parse 
----------------
 []
(1 row)

SELECT simdjson_parse('[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]');  -- OK
                                                                                              simdjson_parse                                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
(1 row)

SELECT simdjson_parse('[1,2]');			-- OK
 simdjson_parse 
----------------
 [1, 2]
(1 row)

SELECT simdjson_parse('[1,2,]');			-- ERROR, trailing comma
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('[1,2');			-- ERROR, no closing bracket
ERROR:  Parsed status: not valid, Problem while parsing a number
SELECT simdjson_parse('[1,[2]');			-- ERROR, no closing bracket
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
-- Objects.
SELECT simdjson_parse('{}');				-- OK
 simdjson_parse 
----------------
 {}
(1 row)

SELECT simdjson_parse('{"abc"}');			-- ERROR, no value
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('{"abc":1}');		-- OK
 simdjson_parse 
----------------
 {"abc": 1}
(1 row)

SELECT simdjson_parse('{1:"abc"}');		-- ERROR, keys must be strings
ERROR:  Parsed status: not valid, Problem while parsing a number
SELECT simdjson_parse('{"abc",1}');		-- ERROR, wrong separator
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('{"abc"=1}');		-- ERROR, totally wrong separator
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('{"abc"::1}');		-- ERROR, another wrong separator
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('{"abc":1,"def":2,"ghi":[3,4],"hij":{"klm":5,"nop":[6]}}'); -- OK
                           simdjson_parse                           
--------------------------------------------------------------------
 {"abc": 1, "def": 2, "ghi": [3, 4], "hij": {"klm": 5, "nop": [6]}}
(1 row)

SELECT simdjson_parse('{"abc":1:2}');		-- ERROR, colon in wrong spot
ERROR:  Parsed status: not valid, Something went wrong while writing to the tape
SELECT simdjson_parse('{"abc":1,3}');		-- ERROR, no value
ERROR:  Parsed status: not valid, Problem while parsing a number
-- Recursion.
SET max_stack_depth = '100kB';
SELECT simdjson_parse(repeat('[', 10000));
ERROR:  Parsed status: not valid, Document exceeds the user-specified depth limitation
SELECT simdjson_parse(repeat('{"a":', 10000));
ERROR:  Parsed status: not valid, Document exceeds the user-specified depth limitation
RESET max_stack_depth;
-- Miscellaneous stuff.
SELECT simdjson_parse('true');			-- OK
 simdjson_parse 
----------------
 true
(1 row)

SELECT simdjson_parse('false');			-- OK
 simdjson_parse 
----------------
 false
(1 row)

SELECT simdjson_parse('null');			-- OK
 simdjson_parse 
----------------
 null
(1 row)

SELECT simdjson_parse(' true ');			-- OK, even with extra whitespace
 simdjson_parse 
----------------
 true
(1 row)

SELECT simdjson_parse('true false');		-- ERROR, too many values
ERROR:  Parsed status: not valid, Problem while parsing an atom starting with the letter 't'
SELECT simdjson_parse('true, false');		-- ERROR, too many values
ERROR:  Parsed status: not valid, Problem while parsing an atom starting with the letter 't'
SELECT simdjson_parse('truf');			-- ERROR, not a keyword
ERROR:  Parsed status: not valid, Problem while parsing an atom starting with the letter 't'
SELECT simdjson_parse('trues');			-- ERROR, not a keyword
ERROR:  Parsed status: not valid, Problem while parsing an atom starting with the letter 't'
SELECT simdjson_parse('');				-- ERROR, no value
ERROR:  Parsed status: not valid, Empty
SELECT simdjson_parse('    ');			-- ERROR, no value
ERROR:  Parsed status: not valid, Empty
